src/ImageSharp/Formats/Qoi/QoiDecoderCore.cs

```diff
@@ -0,0 +1,291 @@
+    /// <exception cref="InvalidImageContentException">If the stream doesn't store a qoi image</exception>
+    private void ProcessHeader(BufferedReadStream stream)
+    {
+        Span<byte> magicBytes = stackalloc byte[4];
+        Span<byte> widthBytes = stackalloc byte[4];
+        Span<byte> heightBytes = stackalloc byte[4];
+
+        // Read magic bytes
+        int read = stream.Read(magicBytes);
+        if (read != 4 || !magicBytes.SequenceEqual(QoiConstants.Magic.ToArray()))
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        // If it's a qoi image, read the rest of properties
+        read = stream.Read(widthBytes);
+        if (read != 4)
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        read = stream.Read(heightBytes);
+        if (read != 4)
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        // These numbers are in Big Endian so we have to reverse them to get the real number
+        uint width = BinaryPrimitives.ReadUInt32BigEndian(widthBytes),
+            height = BinaryPrimitives.ReadUInt32BigEndian(heightBytes);
+        if (width == 0 || height == 0)
+        {
+            throw new InvalidImageContentException(
+                $"The image has an invalid size: width = {width}, height = {height}");
+        }
+
+        int channels = stream.ReadByte();
+        if (channels is -1 or (not 3 and not 4))
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        PixelTypeInfo pixelType = new(8 * channels);
+
+        int colorSpace = stream.ReadByte();
+        if (colorSpace is -1 or (not 0 and not 1))
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        this.header = new QoiHeader(width, height, (QoiChannels)channels, (QoiColorSpace)colorSpace);
+    }
+
+    [DoesNotReturn]
+    private static void ThrowInvalidImageContentException()
+        => throw new InvalidImageContentException("The image is not a valid QOI image.");
+
+    private void ProcessPixels<TPixel>(BufferedReadStream stream, Buffer2D<TPixel> pixels)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        Rgba32[] previouslySeenPixels = new Rgba32[64];
+        Rgba32 previousPixel = new(0, 0, 0, 255);
+
+        // We save the pixel to avoid loosing the fully opaque black pixel
+        // See https://github.com/phoboslab/qoi/issues/258
+        int pixelArrayPosition = GetArrayPosition(previousPixel);
+        previouslySeenPixels[pixelArrayPosition] = previousPixel;
+        byte operationByte;
+        Rgba32 readPixel = default;
+        Span<byte> pixelBytes = MemoryMarshal.CreateSpan(ref Unsafe.As<Rgba32, byte>(ref readPixel), 4);
+        TPixel pixel = default;
+
+        for (int i = 0; i < this.header.Height; i++)
+        {
+            for (int j = 0; j < this.header.Width; j++)
+            {
+                Span<TPixel> row = pixels.DangerousGetRowSpan(i);
```

Comment:
Querying the row span for each pixel actually performs worse than the indexer. I assume you are querying it here and not in the outer loop because of the `QoiOpRun` logic.  Can't you instead "forward" the `row` variable in that branch so you don't penalize the other cases?

Reply:
Oh no no no, my bad, I didn't move it to the correct line